controllers/config/clientconfig_controller.go

```diff
@@ -0,0 +1,206 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	configapi "github.com/k8ssandra/k8ssandra-operator/apis/config/v1beta1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/clientcache"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+)
+
+const (
+	ClientConfigHashAnnotation = k8ssandraapi.ResourceHashAnnotation
+	KubeSecretHashAnnotation   = "k8ssandra.io/secret-hash"
+)
+
+type ClientConfigReconciler struct {
+	Scheme       *runtime.Scheme
+	ClientCache  *clientcache.ClientCache
+	shutdownFunc context.CancelFunc
+
+	// filterMutex  sync.RWMutex
+	secretFilter map[types.NamespacedName]types.NamespacedName
+}
+
+/*
+	What about situations:
+		- Remove clientconfig when the cluster is still added?
+			* Cluster properties get updated?
+				* Do we accept or not?
+			* Could just be a cert-update etc, short-term, long-term ones
+		- New ClientConfig doesn't work?
+		- Add cluster points to a clientconfig that doesn't exist?
+		- What if the ClientConfig's kubeConfig is changed in the secret?
+		- What if certs expire? What if client connection fails - how do we detect when we need to do something to clientConfig?
```

Comment:
Excellent questions but outside the scope of what we trying to do here. It would be great to create issues for these. 

Cert expiration and connection failing are distinct and I think should be handled separately. Cert expiration is something we should be able to determine easily, failed client connections not so much. How do we distinguish between transitory failures and a permanently bad connection?

Reply:
Actually, k8ssandra-operator fails to start at the moment if it can't connect to the cluster it was configured to connect to. The kubeConfig -> RESTClient part throws an error and the initialization fails in main.go